docs/testing.org

```diff
@@ -0,0 +1,274 @@
+running scripts against the hammerspoon CLI =hs=.
+
+Tests can be ran invoking the following shell command within the =~/.hammerspoon= directory:
+
+#+begin_src bash :dir ..
+./run-test test/*.fnl
+#+end_src
+
+Which will output something like the following:
+
+#+begin_example
+Running tests for       /Users/j/.hammerspoon/test/functional-test.fnl
+Running tests for       /Users/j/.hammerspoon/test/statemachine-test.fnl
+
+Functional
+
+  Call when calls function if it exists ...
+    [ OK ]
+
+  Compose combines functions together in reverse order ...
+    [ OK ]
+
+  Contains? returns true if list table contains a value ...
+    [ OK ]
+
+State Machine
+
+  Should create a new fsm in the closed state ...
+    [ OK ]
+
+  Should transition to opened on open event ...
+    [ OK ]
+
+  Should transition back to opened on close event ...
+    [ OK ]
+
+  Should not explode when dispatching an unhandled event ...
+05:27:10 ** Warning:    statemach: Could not fail from closed state
+    [ OK ]
+
+
+  Ran 7 tests 7 passed 0 failed in 0.038301000000047 seconds
+#+end_example
+
+* Requirements
+
+Be sure to run the =hs.ipc.cliInstall= [[https://www.hammerspoon.org/docs/hs.ipc.html#cliInstall][command]] from hammerspoon. You may paste
+this or eval against the Hammerspoon console:
+
+#+begin_src lua
+hs.ipc.cliInstall()
+#+end_src
+
+* Testing API
+
+The current form of the testing API is inspired by JS libraries like [[https://mochajs.org/][mocha]] given
+how easy it is to implement.
+
+** Describe
+
+Label a test suite
+
+Usage:
+
+#+begin_src fennel
+(describe
+ "Functional Tools"
+ (fn []
+   ;; Other describe calls or `it` tests can run here
+   )
+#+end_src
+
+Describe a suite of tests contained in its function body. The function
+body may contain other =describe= calls as well as =it= calls to perform tests.
+
+The aim is to help organize displayed test results, as its inner tests
+are indented underneath the describe text label when printing test results.
+
+** It
+
+Perform a test that can either pass or fail.
+
+Usage:
+
+#+begin_src fennel
+(describe
+ "Basic Fennel Tests"
+ (fn []
+   (it "Should do math"
+       (fn []
+         (is.eq? (+ 1 1) 2 "Did not result in 2")))))
+#+end_src
+
+The bodies of =it= calls should run code and perform assertions, if no
+error is thrown, the test has passed.
+
+=it= calls cannot be nested, instead should have siblings within a
+=describe= suite.
+
+** Before
+
+Run a function before tests run in a suite
+
+Usage:
+
+#+begin_src fennel
+(describe
+ "Functional Tools"
+ (fn []
+   (before (fn []
+             (print "Perform pre-test setup")))
+
+   (it "Should do math"
+       (fn []
+         (is.eq? (+ 1 1) 2 "Did not result in 2")))))
+#+end_src
+
+=before= is best used as a way to prepare data, or allocate resources
+tests may use before they're setup
+
+*** Does =before= run before each =it=?
+
+Before runs before all =it= calls within a describe block. If you wanted
```

Comment:
for extra clarity, something like: Before runs _once_ before **all** `it` calls within a..